ActivityAnalyses/gait_analysis.py

```diff
@@ -732,6 +761,17 @@ def detect_correct_order(rHS, rTO, lHS, lTO):
         l_toe_rel_x = (
             self.markerDict['markers']['L_toe_study'] - 
             self.markerDict['markers']['L.PSIS_study'])[:,0]
+        
+        # Identify which direction the subject is walking.
```

Comment:
Good catch. I suggest we go even more general:
1) find heading vector from mid-psis to mid-asis; project onto floor (ie make y=0)
2) dot projected heading vector with the l_calc_rel vecor (get rid of the [:,0] before dot.
3) the result should function how the r_toe_rel_x, etc... is already functioning.

Reply:
This will help if they walk off axis from the checkerboard at all

Reply:
@suhlrich I am not super clear on what you wrote. Is this what you mean?

```
      r_calc_rel = (
          self.markerDict['markers']['r_calc_study'] - 
          self.markerDict['markers']['r.PSIS_study'])
      r_calc_rel_x = r_calc_rel[:,0]
      r_toe_rel = (
          self.markerDict['markers']['r_toe_study'] - 
          self.markerDict['markers']['r.PSIS_study'])
      r_toe_rel_x = r_toe_rel[:,0]

      # Repeat for left.
      l_calc_rel = (
          self.markerDict['markers']['L_calc_study'] - 
          self.markerDict['markers']['L.PSIS_study'])
      l_calc_rel_x = l_calc_rel[:,0]
      l_toe_rel = (
          self.markerDict['markers']['L_toe_study'] - 
          self.markerDict['markers']['L.PSIS_study'])
      l_toe_rel_x = l_toe_rel[:,0]
      
      # Identify which direction the subject is walking.
      mid_psis = (self.markerDict['markers']['r.PSIS_study'] + self.markerDict['markers']['L.PSIS_study'])/2
      mid_asis = (self.markerDict['markers']['r.ASIS_study'] + self.markerDict['markers']['L.ASIS_study'])/2
      mid_dir = mid_asis - mid_psis
      mid_dir_floor = np.copy(mid_dir)
      mid_dir_floor[:,1] = 0
      mid_dir_floor = mid_dir_floor / np.linalg.norm(mid_dir_floor,axis=1,keepdims=True)
      l_calc_rel_floor = np.copy(l_calc_rel)
      l_calc_rel_floor[:,1] = 0
      l_calc_rel_floor = l_calc_rel_floor / np.linalg.norm(l_calc_rel_floor,axis=1,keepdims=True)        
      dot_product = np.einsum('ij,ij->i', mid_dir_floor,l_calc_rel_floor)
```

How do you then use the dot_product?

Reply:
`      r_calc_rel = (
          self.markerDict['markers']['r_calc_study'] - 
          self.markerDict['markers']['r.PSIS_study'])

      r_toe_rel = (
          self.markerDict['markers']['r_toe_study'] - 
          self.markerDict['markers']['r.PSIS_study'])


      # Repeat for left.
      l_calc_rel = (
          self.markerDict['markers']['L_calc_study'] - 
          self.markerDict['markers']['L.PSIS_study'])

      l_toe_rel = (
          self.markerDict['markers']['L_toe_study'] - 
          self.markerDict['markers']['L.PSIS_study'])

      
      # Identify which direction the subject is walking.
      mid_psis = (self.markerDict['markers']['r.PSIS_study'] + self.markerDict['markers']['L.PSIS_study'])/2
      mid_asis = (self.markerDict['markers']['r.ASIS_study'] + self.markerDict['markers']['L.ASIS_study'])/2
      mid_dir = mid_asis - mid_psis
      mid_dir_floor = np.copy(mid_dir)
      mid_dir_floor[:,1] = 0
      mid_dir_floor = mid_dir_floor / np.linalg.norm(mid_dir_floor,axis=1,keepdims=True)
      l_calc_rel_floor = np.copy(l_calc_rel)
      l_calc_rel_floor[:,1] = 0
      l_calc_rel_floor = l_calc_rel_floor / np.linalg.norm(l_calc_rel_floor,axis=1,keepdims=True)        
      l_calc_rel_x = np.einsum('ij,ij->i', mid_dir_floor,l_calc_rel_floor)
      
      # TODO RIGHT SIDE
      
      `

Reply:
Ah I see, I adjusted. Checked the results on a few cases and very very small differences (like one frame difference causing a 1% change in symmetry). I also visualized old and new l_calc_rel_x  etc and they are essentially on top of each other, so I think we are good. Another look at the PR would be great though. Thanks.